File: rvlib/rvlib/render.py
import os
import sys
import argparse
import bpy

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
if _THIS_DIR not in sys.path:
    sys.path.insert(0, _THIS_DIR)

from runtime_bootstrap import bootstrap_runtime  # noqa: E402


def parse_args():
    args = []
    if "--" in sys.argv:
        args = sys.argv[sys.argv.index("--") + 1 :]

    parser = argparse.ArgumentParser(description="Render runner")
    parser.add_argument("--script", type=str)
    parser.add_argument("--libpath", type=str)
    parser.add_argument("--output", type=str)
    parser.add_argument("--number", type=int)
    parser.add_argument("--resolution", type=str, default="640,640")
    parser.add_argument("--gpu-backend", type=str, default="auto")
    parser.add_argument("--seed-mode", type=str, default="rand")
    parser.add_argument("--seed-value", type=int, default=None)
    parser.add_argument("--seed-base", type=int, default=0)
    parser.add_argument("--time-limit", type=float, default=None)
    parser.add_argument("--max-samples", type=int, default=None)
    parser.add_argument("--min-samples", type=int, default=None)
    parser.add_argument(
        "--noise-threshold-enabled",
        action=argparse.BooleanOptionalAction,
        default=None,
    )
    parser.add_argument("--noise-threshold", type=float, default=None)
    parser.add_argument("--cwd", type=str)
    parser.add_argument("--generator-port", type=int, default=0)

    return parser.parse_args(args)


def apply_cycles_overrides(
    max_samples, min_samples, noise_threshold_enabled, noise_threshold
):
    scene_cycles = bpy.context.scene.cycles

    if max_samples is not None:
        if max_samples <= 0:
            raise ValueError("--max-samples must be > 0")
        if not hasattr(scene_cycles, "samples"):
            raise AttributeError("Blender cycles.samples is not available")
        scene_cycles.samples = max_samples

    if min_samples is not None:
        if min_samples < 0:
            raise ValueError("--min-samples must be >= 0")
        if not hasattr(scene_cycles, "adaptive_min_samples"):
            raise AttributeError("Blender cycles.adaptive_min_samples is not available")
        scene_cycles.adaptive_min_samples = min_samples

    if min_samples is not None and max_samples is not None and min_samples > max_samples:
        raise ValueError("--min-samples must be <= --max-samples")

    if noise_threshold is not None:
        if noise_threshold_enabled is not True:
            raise ValueError("--noise-threshold requires --noise-threshold-enabled=true")
        if noise_threshold <= 0:
            raise ValueError(
                "--noise-threshold must be > 0 when --noise-threshold-enabled=true"
            )

    if noise_threshold_enabled is not None:
        if not hasattr(scene_cycles, "use_adaptive_sampling"):
            raise AttributeError("Blender cycles.use_adaptive_sampling is not available")
        scene_cycles.use_adaptive_sampling = noise_threshold_enabled

    if noise_threshold_enabled is True:
        if noise_threshold is None:
            raise ValueError(
                "--noise-threshold is required when --noise-threshold-enabled=true"
            )
        if not hasattr(scene_cycles, "adaptive_threshold"):
            raise AttributeError("Blender cycles.adaptive_threshold is not available")
        scene_cycles.adaptive_threshold = noise_threshold


def run_script(
    script_path,
    output_dir,
    resolution,
    gpu_backend,
    time_limit,
    max_samples,
    min_samples,
    noise_threshold_enabled,
    noise_threshold,
    seed,
):
    import rv.internal as rvi

    scene_class = rvi._internal_load_scene_class(script_path)

    def execute_run():
        rvi._internal_begin_run(purge_orphans=True)
        instance = scene_class(output_dir)
        instance.resolution = resolution
        rvi._internal_set_time_limit(instance, time_limit)
        rvi._internal_run_scene_generate(instance, seed, ARGS.seed_mode)
        instance._internal_post_gen()
        apply_cycles_overrides(
            max_samples=max_samples,
            min_samples=min_samples,
            noise_threshold_enabled=noise_threshold_enabled,
            noise_threshold=noise_threshold,
        )
        selected_backend = rvi._internal_configure_cycles_backend(gpu_backend)
        print(f"[rv] selected_gpu_backend={selected_backend}")
        rvi._internal_print_cycles_device_info()
        instance._internal_render()
        instance._internal_save_metadata("_meta.json")
        rvi._internal_end_run(purge_orphans=False)

    execute_run()


ARGS = parse_args()

if ARGS.libpath is None:
    raise ValueError("--libpath is required")
if ARGS.cwd is None:
    raise ValueError("--cwd is required")
if ARGS.number is None:
    raise ValueError("--number is required")
if ARGS.script is None:
    raise ValueError("--script is required")

bootstrap_runtime(ARGS.libpath, ARGS.cwd)

import rv.internal as rvi  # noqa: E402

RESOLUTION = rvi._internal_parse_resolution(ARGS.resolution)
rvi._configure_generator_runtime(ARGS.generator_port, ARGS.cwd)

for i in range(ARGS.number):
    seed = rvi._internal_resolve_seed(
        ARGS.seed_mode, ARGS.seed_value, ARGS.seed_base, i
    )
    run_script(
        ARGS.script,
        ARGS.output,
        RESOLUTION,
        ARGS.gpu_backend,
        ARGS.time_limit,
        ARGS.max_samples,
        ARGS.min_samples,
        ARGS.noise_threshold_enabled,
        ARGS.noise_threshold,
        seed,
    )

File: internal/render/render.go
package render

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"slices"
	"strings"
	"syscall"

	"github.com/Rapid-Vision/rv/internal/generator"
	"github.com/Rapid-Vision/rv/internal/logs"
	"github.com/Rapid-Vision/rv/internal/seed"
	"github.com/Rapid-Vision/rv/internal/utils"
)

type RenderOptions struct {
	ScriptPath            string
	Cwd                   string
	GenBaseDir            string
	WorkDir               string
	GenRetain             utils.GeneratorRetention
	ImageNum              int
	Procs                 int
	Resolution            [2]int
	OutputDir             string
	GPUBackend            string
	TimeLimit             *float64
	MaxSamples            *int
	MinSamples            *int
	NoiseThresholdEnabled *bool
	NoiseThreshold        *float64
	Seed                  seed.Config
	GeneratorPort         int
}

type RenderResult struct {
	OutputDir string
}

type datasetMeta struct {
	Resolution   [2]int             `json:"resolution"`
	RenderParams datasetRenderParms `json:"render_params"`
	SampleFiles  []string           `json:"sample_files"`
}

type datasetRenderParms struct {
	ScriptPath            string   `json:"script_path"`
	Cwd                   string   `json:"cwd"`
	Number                int      `json:"number"`
	Procs                 int      `json:"procs"`
	GPUBackend            string   `json:"gpu_backend,omitempty"`
	TimeLimit             *float64 `json:"time_limit,omitempty"`
	MaxSamples            *int     `json:"max_samples,omitempty"`
	MinSamples            *int     `json:"min_samples,omitempty"`
	NoiseThresholdEnabled *bool    `json:"noise_threshold_enabled,omitempty"`
	NoiseThreshold        *float64 `json:"noise_threshold,omitempty"`
	SeedMode              string   `json:"seed_mode,omitempty"`
	SeedValue             *int64   `json:"seed_value,omitempty"`
}

func normalizedGPUBackend(value string) string {
	if value == "" {
		return "auto"
	}
	return strings.ToLower(value)
}

func applyDefaultGeneratorOptions(opts *RenderOptions) error {
	if opts.GenBaseDir == "" {
		generatorPaths, err := utils.ResolveGeneratorPaths(opts.Cwd, "")
		if err != nil {
			return fmt.Errorf("resolve generator paths: %w", err)
		}
		opts.GenBaseDir = generatorPaths.GenBaseDir
	}
	if opts.GenRetain == "" {
		opts.GenRetain = utils.GeneratorRetainNone
	}
	return nil
}

func Render(opts RenderOptions) (RenderResult, error) {
	opts.Seed = seed.Normalize(opts.Seed)

	scriptPath := opts.ScriptPath
	imgNum := opts.ImageNum
	procs := opts.Procs
	resolution := opts.Resolution
	outputDir := opts.OutputDir
	cwdAbs := opts.Cwd

	if scriptPath == "" {
		return RenderResult{}, errors.New("script path is required")
	}
	if outputDir == "" {
		return RenderResult{}, errors.New("output directory is required")
	}
	if cwdAbs == "" {
		return RenderResult{}, errors.New("cwd is required")
	}
	if err := applyDefaultGeneratorOptions(&opts); err != nil {
		return RenderResult{}, err
	}

	blenderPath, err := utils.GetBlenderPath()
	if err != nil {
		return RenderResult{}, fmt.Errorf("can't find blender path: %w", err)
	}

	libPath, err := utils.GetLibPath()
	if err != nil {
		return RenderResult{}, fmt.Errorf("can't find rvlib: %w", err)
	}

	logs.Info.Println("librv path: ", libPath)

	generatorCtx, cancelGenerator := context.WithCancel(context.Background())
	generatorService, err := generator.Start(generatorCtx)
	if err != nil {
		cancelGenerator()
		logs.Warn.Printf("Generator service unavailable; scenes using self.generators will fail: %v\n", err)
		opts.GeneratorPort = 0
	} else {
		opts.GeneratorPort = generatorService.Port()
		defer func() {
			cancelGenerator()
			generatorService.Wait()
		}()
	}

	seqOutDir, err := utils.GetSequentialOutputDir(outputDir)
	if err != nil {
		return RenderResult{}, fmt.Errorf("can't create new output directory: %w", err)
	}

	var cmdBuff [](*exec.Cmd)

	if imgNum < 1 {
		return RenderResult{}, errors.New("--number is less then one")
	}

	if procs < 1 {
		return RenderResult{}, errors.New("--procs must be at least 1")
	}
	if resolution[0] <= 0 || resolution[1] <= 0 {
		return RenderResult{}, errors.New("--resolution must be WIDTH,HEIGHT with positive integers")
	}
	opts.GPUBackend = normalizedGPUBackend(opts.GPUBackend)
	if err := validateOptionalRenderOptions(opts); err != nil {
		return RenderResult{}, err
	}

	workDir, err := utils.AllocateGeneratorWorkDir(opts.GenBaseDir)
	if err != nil {
		return RenderResult{}, err
	}
	opts.WorkDir = workDir
	if opts.GenRetain == utils.GeneratorRetainLast || opts.GenRetain == utils.GeneratorRetainNone {
		if err := utils.CleanupGeneratorWorkDirs(opts.GenBaseDir, utils.GeneratorRetainLast, opts.WorkDir); err != nil {
			return RenderResult{}, err
		}
	}
	defer func() {
		if err := utils.CleanupGeneratorWorkDirs(opts.GenBaseDir, opts.GenRetain, opts.WorkDir); err != nil {
			logs.Warn.Printf("Failed to clean generator work directories: %v\n", err)
		}
	}()

	if imgNum < procs {
		procs = imgNum
	}

	for i := 0; i < procs; i += 1 {
		part := utils.SplitTaskBetweenProcs(imgNum, procs, i)
		seedBase := renderSeedBase(imgNum, procs, i)

		// Start Blender
		cmd := exec.Command(
			blenderPath,
			buildBlenderRenderArgs(opts, libPath, seqOutDir, part, seedBase)...,
		)
		cmd.Env = utils.BlenderCommandEnv()
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr

		if err = cmd.Start(); err != nil {
			return RenderResult{}, fmt.Errorf("failed to start blender: %w", err)
		}
		logs.Info.Printf("Blender started (PID %d)\n", cmd.Process.Pid)

		cmdBuff = append(cmdBuff, cmd)
	}

	// Handle Ctrl-C
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM)

	waitCh := utils.WaitCmdBuff(cmdBuff)
	var renderErr error

	// Wait for either all Blender instances exit or signal
	for i := 0; i < procs; i += 1 {
		select {
		case <-sigCh:
			logs.Info.Println("Interrupt received — terminating Blender…")

			for j := 0; j < procs; j += 1 {
				_ = cmdBuff[j].Process.Signal(syscall.SIGTERM)
			}
		case err = <-waitCh:
			if err != nil {
				logs.Info.Println("Blender exited with error:", err)
				if renderErr == nil {
					renderErr = err
				}
			} else {
				logs.Info.Println("Blender exited.")
			}
		}
	}

	if renderErr != nil {
		return RenderResult{}, renderErr
	}
	if err := writeDatasetMetadata(seqOutDir, opts); err != nil {
		return RenderResult{}, fmt.Errorf("failed to write dataset metadata: %w", err)
	}

	return RenderResult{
		OutputDir: seqOutDir,
	}, nil
}

func buildBlenderRenderArgs(opts RenderOptions, libPath string, seqOutDir string, part int, seedBase int) []string {
	opts.Seed = seed.Normalize(opts.Seed)
	gpuBackend := normalizedGPUBackend(opts.GPUBackend)
	args := []string{
		filepath.Join(libPath, "template.blend"),
		"--factory-startup",
		"--background",
		"--python-exit-code", "1",
		"--python", filepath.Join(libPath, "render.py"),
		"--",
		"--script", opts.ScriptPath,
		"--libpath", libPath,
		"--number", fmt.Sprintf("%d", part),
		"--resolution", fmt.Sprintf("%d,%d", opts.Resolution[0], opts.Resolution[1]),
		"--output", seqOutDir,
		"--root-dir", opts.Cwd,
		"--work-dir", opts.WorkDir,
		"--gpu-backend", gpuBackend,
		"--seed-mode", string(opts.Seed.Mode),
		"--seed-base", fmt.Sprintf("%d", seedBase),
		"--generator-port", fmt.Sprintf("%d", opts.GeneratorPort),
	}
	if opts.Seed.Mode == seed.FixedMode {
		args = append(args, "--seed-value", fmt.Sprintf("%d", opts.Seed.Value))
	}

	if opts.TimeLimit != nil {
		args = append(args, "--time-limit", fmt.Sprintf("%g", *opts.TimeLimit))
	}
	if opts.MaxSamples != nil {
		args = append(args, "--max-samples", fmt.Sprintf("%d", *opts.MaxSamples))
	}
	if opts.MinSamples != nil {
		args = append(args, "--min-samples", fmt.Sprintf("%d", *opts.MinSamples))
	}
	if opts.NoiseThresholdEnabled != nil {
		if *opts.NoiseThresholdEnabled {
			args = append(args, "--noise-threshold-enabled")
		} else {
			args = append(args, "--no-noise-threshold-enabled")
		}
	}
	if opts.NoiseThreshold != nil {
		args = append(args, "--noise-threshold", fmt.Sprintf("%g", *opts.NoiseThreshold))
	}

	return args
}

func renderSeedBase(imageNum int, procs int, procIndex int) int {
	base := 0
	for i := range procIndex {
		base += utils.SplitTaskBetweenProcs(imageNum, procs, i)
	}
	return base
}

func validateOptionalRenderOptions(opts RenderOptions) error {
	switch normalizedGPUBackend(opts.GPUBackend) {
	case "auto", "optix", "cuda", "hip", "oneapi", "metal", "cpu":
	default:
		return errors.New("--gpu-backend must be one of auto, optix, cuda, hip, oneapi, metal, cpu")
	}
	switch opts.Seed.Mode {
	case "", seed.RandomMode, seed.SeqMode, seed.FixedMode:
	default:
		return errors.New("invalid seed mode")
	}
	if opts.TimeLimit != nil && *opts.TimeLimit <= 0 {
		return errors.New("--time-limit must be > 0")
	}
	if opts.MaxSamples != nil && *opts.MaxSamples <= 0 {
		return errors.New("--max-samples must be > 0")
	}
	if opts.MinSamples != nil && *opts.MinSamples < 0 {
		return errors.New("--min-samples must be >= 0")
	}
	if opts.MinSamples != nil && opts.MaxSamples != nil && *opts.MinSamples > *opts.MaxSamples {
		return errors.New("--min-samples must be <= --max-samples")
	}
	if opts.NoiseThresholdEnabled != nil && *opts.NoiseThresholdEnabled {
		if opts.NoiseThreshold == nil {
			return errors.New("--noise-threshold is required when --noise-threshold-enabled=true")
		}
		if *opts.NoiseThreshold <= 0 {
			return errors.New("--noise-threshold must be > 0 when --noise-threshold-enabled=true")
		}
	}
	if opts.NoiseThreshold != nil {
		if opts.NoiseThresholdEnabled == nil || !*opts.NoiseThresholdEnabled {
			return errors.New("--noise-threshold requires --noise-threshold-enabled=true")
		}
	}
	return nil
}

func writeDatasetMetadata(seqOutDir string, opts RenderOptions) error {
	if err := os.MkdirAll(seqOutDir, os.ModePerm); err != nil {
		return err
	}

	entries, err := os.ReadDir(seqOutDir)
	if err != nil {
		return err
	}

	samples := make([]string, 0)
	sampleFiles := make([]string, 0)
	haveSampleFiles := false
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}

		samplePath := filepath.Join(seqOutDir, entry.Name())
		if !haveSampleFiles {
			files, err := listFilesRecursive(samplePath)
			if err != nil {
				return err
			}
			sampleFiles = files
			haveSampleFiles = true
		}
		samples = append(samples, entry.Name())
	}
	slices.SortFunc(samples, strings.Compare)

	meta := datasetMeta{
		Resolution: opts.Resolution,
		RenderParams: datasetRenderParms{
			ScriptPath:            opts.ScriptPath,
			Cwd:                   opts.Cwd,
			Number:                opts.ImageNum,
			Procs:                 opts.Procs,
			GPUBackend:            normalizedGPUBackend(opts.GPUBackend),
			TimeLimit:             opts.TimeLimit,
			MaxSamples:            opts.MaxSamples,
			MinSamples:            opts.MinSamples,
			NoiseThresholdEnabled: opts.NoiseThresholdEnabled,
			NoiseThreshold:        opts.NoiseThreshold,
			SeedMode:              string(seed.Normalize(opts.Seed).Mode),
		},
		SampleFiles: sampleFiles,
	}
	normalizedSeed := seed.Normalize(opts.Seed)
	if normalizedSeed.Mode == seed.FixedMode {
		meta.RenderParams.SeedValue = &normalizedSeed.Value
	}

	f, err := os.Create(filepath.Join(seqOutDir, "_meta.json"))
	if err != nil {
		return err
	}
	defer func() { _ = f.Close() }()

	enc := json.NewEncoder(f)
	enc.SetIndent("", "    ")
	return enc.Encode(meta)
}

func listFilesRecursive(root string) ([]string, error) {
	files := make([]string, 0)
	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(root, path)
		if err != nil {
			return err
		}
		files = append(files, filepath.ToSlash(rel))
		return nil
	})
	if err != nil {
		return nil, err
	}
	slices.Sort(files)
	return files, nil
}
